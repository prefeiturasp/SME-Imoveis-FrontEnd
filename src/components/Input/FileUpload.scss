.input.input-file-upload {
  font-size: 10px;
}
