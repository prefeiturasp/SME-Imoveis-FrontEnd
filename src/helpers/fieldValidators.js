export const required = value =>
  value !== undefined ? undefined : "Campo obrigatório";

export const email = value =>
  value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value)
    ? "Email inválido"
    : undefined;

export const prefeituraEmail = value =>
  value && /.+@\prefeitura.sp.gov.br/.test(value)
    ? undefined
    : "Somente emails da prefeitura de São Paulo";

export const alphaNumeric = value =>
  value && /[^a-zA-Z0-9 ]/i.test(value)
    ? "Only alphanumeric characters"
    : undefined;

export const numericInteger = value =>
  value && /[^0-9 ]/i.test(value) ? "Somente números" : undefined;

export const phoneNumber = value =>
  value && !/^(0|[1-9][0-9]{9})$/i.test(value)
    ? "Invalid phone number, must be 10 digits"
    : undefined;

export const minLength = min => value =>
  value && value.length < min
    ? `Deve ter ao menos ${min} caracteres(s)`
    : undefined;

export const telValidate = value => {
  if (value) {
    const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
    if (cleanedValue.length < 10 || cleanedValue.length > 11)
      return "Telefone inválido";
    return undefined;
  }
  return undefined;
};

export const celValidate = value => {
  if (value) {
    const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
    if (cleanedValue.length !== 11) return "Celular inválido";
    return undefined;
  }
  return undefined;
};

export const nameValidate = value =>
  value && !/^[a-zA-Z]+. +[a-zA-Z]/.test(value) ? "Nome inválido" : undefined;

export const composeValidators = (...validators) => value =>
  validators.reduce((error, validator) => error || validator(value), undefined);
