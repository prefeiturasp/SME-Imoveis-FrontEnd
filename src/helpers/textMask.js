import { createTextMask } from "redux-form-input-masks";
import formatString from "format-string-by-pattern";

export const fieldCPF_CNPJ = value => {
  const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
  if (cleanedValue.length <= 11) {
    return formatString("999.999.999-99", value);
  } else {
    return formatString("99.999.999/9999-99", value);
  }
};

export const fieldCep = createTextMask({
  pattern: "99999-999",
  allowEmpty: false,
  guide: true,
  stripMask: false
});

export const fieldTel = createTextMask({
  pattern: "(99) 99999-9999",
  allowEmpty: false,
  guide: true,
  stripMask: false
});

export const fieldCel = createTextMask({
  pattern: "(99) 99999-9999",
  allowEmpty: false,
  guide: true,
  stripMask: false
});

export const telCelMask = value => {
  const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
  if (cleanedValue.length <= 10) {
    return formatString("(99) 9999-9999", value);
  } else {
    return formatString("(99) 9 9999-9999", value);
  }
};

export const codigoEscolaMask = value => {
  const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
  return formatString("000000", cleanedValue);
};

export const iptuMask = value => {
  const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
  return formatString("999.999.9999.9", cleanedValue);
};


export const processoSeiMask = value => {
  const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
  return formatString("9999.9999/9999999-9", cleanedValue);
};

export const numeroProtocolo = value => {
  const cleanedValue = value.replace(/[^a-z0-9]/gi, "").replace(/\D/g, "");
  return formatString("000/00", cleanedValue);
};