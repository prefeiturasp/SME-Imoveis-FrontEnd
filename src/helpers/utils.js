export async function asyncForEach(array, callback) {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
}

export async function readerFile(file) {
  let result_file = await new Promise((resolve) => {
    const reader = new FileReader();
    reader.onload = () => {
      const re = /(?:\.([^.]+))?$/;
      const base64 = reader.result.split("base64,")[1];
      return resolve({
        arquivo: `data:${file.type}/${re.exec(file.name)[1]};base64,${base64}`,
      });
    };
    reader.readAsDataURL(file);
  });
  return result_file;
}

export const getKey = (obj) => {
  return Object.keys(obj)[0];
};

export const getError = (obj) => {
  let result = "Erro ao cadastrar Imóvel";
  if (!obj[getKey(obj)]) {
    return "Erro ao cadastrar Imóvel";
  } else if (
    (obj[getKey(obj)][0] !== undefined &&
      typeof obj[getKey(obj)][0] !== "string") ||
    typeof obj[getKey(obj)] !== "string"
  ) {
    result = getError(obj[getKey(obj)]);
  } else {
    if (typeof obj[getKey(obj)] === "string") return obj[getKey(obj)];
    else return obj[getKey(obj)][0];
  }
  return result;
};

export const validarCPF = (cpf) => {
  cpf = cpf.replace(/[^\d]+/g, "");
  if (cpf === "") return false;
  // Elimina CPFs invalidos conhecidos
  if (
    cpf.length !== 11 ||
    cpf === "00000000000" ||
    cpf === "11111111111" ||
    cpf === "22222222222" ||
    cpf === "33333333333" ||
    cpf === "44444444444" ||
    cpf === "55555555555" ||
    cpf === "66666666666" ||
    cpf === "77777777777" ||
    cpf === "88888888888" ||
    cpf === "99999999999"
  )
    return false;
  // Valida 1o digito
  let add = 0;
  let i, rev;
  for (i = 0; i < 9; i++) add += parseInt(cpf.charAt(i)) * (10 - i);
  rev = 11 - (add % 11);
  if (rev === 10 || rev === 11) rev = 0;
  if (rev !== parseInt(cpf.charAt(9))) return false;
  // Valida 2o digito
  add = 0;
  for (i = 0; i < 10; i++) add += parseInt(cpf.charAt(i)) * (11 - i);
  rev = 11 - (add % 11);
  if (rev === 10 || rev === 11) rev = 0;
  if (rev !== parseInt(cpf.charAt(10))) return false;
  return true;
};

export const validarCNPJ = (cnpj) => {
  cnpj = cnpj.replace(/[^\d]+/g, "");

  if (cnpj === "") return false;

  if (cnpj.length !== 14) return false;

  // Elimina CNPJs invalidos conhecidos
  if (
    cnpj === "00000000000000" ||
    cnpj === "11111111111111" ||
    cnpj === "22222222222222" ||
    cnpj === "33333333333333" ||
    cnpj === "44444444444444" ||
    cnpj === "55555555555555" ||
    cnpj === "66666666666666" ||
    cnpj === "77777777777777" ||
    cnpj === "88888888888888" ||
    cnpj === "99999999999999"
  )
    return false;

  // Valida DVs
  let i;
  let tamanho = cnpj.length - 2;
  let numeros = cnpj.substring(0, tamanho);
  let digitos = cnpj.substring(tamanho);
  let soma = 0;
  let pos = tamanho - 7;
  for (i = tamanho; i >= 1; i--) {
    soma += numeros.charAt(tamanho - i) * pos--;
    if (pos < 2) pos = 9;
  }
  let resultado = soma % 11 < 2 ? 0 : 11 - (soma % 11);
  if (resultado !== parseInt(digitos.charAt(0))) return false;

  tamanho = tamanho + 1;
  numeros = cnpj.substring(0, tamanho);
  soma = 0;
  pos = tamanho - 7;
  for (i = tamanho; i >= 1; i--) {
    soma += numeros.charAt(tamanho - i) * pos--;
    if (pos < 2) pos = 9;
  }
  resultado = soma % 11 < 2 ? 0 : 11 - (soma % 11);
  if (resultado !== parseInt(digitos.charAt(1))) return false;

  return true;
};

export const hasNumber = (myString) => {
  return /\d/.test(myString);
};

export const truncarString = (str, numeroMaximoChars) => {
  if (str.length > numeroMaximoChars) {
    return str.slice(0, numeroMaximoChars) + "...";
  } else {
    return str;
  }
};

export const tamanhoMaximoAnexos = (anexos, tamanho) => {
  let tamanhoAnexos = 0;
  anexos.forEach((anexo) => {
    tamanhoAnexos += anexo.size;
  });
  if (tamanhoAnexos >= tamanho) return true;
  return false;
};

export const getNome = () => {
  return localStorage.getItem("nome");
};

export const getPerfil = () => {
  return localStorage.getItem("perfil");
};

export const normalizarOptions = (options) => {
  const options_ = [{ label: "Selecione", value: undefined }];
  options.forEach((option) => {
    options_.push({ label: option.nome, value: option.id });
  });
  return options_;
};

export const normalizarSetores = (options) => {
  const options_ = [{ label: "Selecione", value: undefined }];
  options.forEach((option) => {
    options_.push({ label: option.codigo, value: option.codigo });
  });
  return options_;
};


export const normalizarPerfis = (options) => {
  const options_ = [
    { label: "Selecione", value: undefined }
  ];
  options.forEach((option) => {
    options_.push({ label: option.nome, value: option.id });
  });
  options_.push({ label: "SEM PERMISSAO", value: "SEM PERMISSAO" })
  return options_;
};

export const EH_PERFIL_ADMIN = localStorage.getItem("perfil") === "ADMIN";
export const EH_PERFIL_SECRETARIA = localStorage.getItem("perfil") === "SECRETARIA";
export const EH_PERFIL_DRE = localStorage.getItem("perfil") === "DRE";
export const EH_PERFIL_CONSULTA_SECRETARIA = localStorage.getItem("perfil") === "CONSULTA_SECRETARIA";