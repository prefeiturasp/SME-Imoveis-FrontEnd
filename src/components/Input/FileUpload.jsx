import React from "react";
import { FileUpload as FileUploadPR } from "primereact/fileupload";

import { InputErroMensagem } from "./InputErroMensagem";
import { HelpText } from "components/HelpText";
import { asyncForEach, readerFile } from "helpers/utils";
import { OverlayTrigger, Popover, Image } from "react-bootstrap";
import imgTooltipQuestion from "img/tooltip-question.svg";

import "./FileUpload.scss";

class CustomFileUploadPR extends FileUploadPR {
  async componentDidMount() {
    const { initialFiles } = this.props;
    if (initialFiles) {
      this.setState(
        {
          files: initialFiles
        },
        this.upload
      );
    }
  }
  async upload() {
    const { onUploadChange, tipoDocumento, tipoArquivo } = this.props;
    const { files } = this.state;
    let data = [];

    await asyncForEach(files, async file => {
      await readerFile(file).then(v => {
        data.push({
          ...v,
          originalFile: file,
          tipo_documento: tipoDocumento,
          tipo_arquivo: tipoArquivo
        });
      });
    });
    onUploadChange(data);
  }
  async remove(index) {
    this.clearInputElement();

    const currentFiles = this.state.files.filter((v, i) => {
      return i !== index;
    });

    this.setState(
      {
        files: currentFiles
      },
      this.upload
    );
  }
}

export class FileUpload extends React.Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.fileUpload = React.createRef();
  }

  onChange(data) {
    const {
      input: { onChange }
    } = this.props;
    if (data) {
      onChange(data);
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.resetarFile && !prevProps.resetarFile) {
      this.fileUpload.current.clear();
      this.props.setResetarFileFalse();
    }
  }

  popover = msg => (
    <Popover id="popover-basic">
      <Popover.Content>{msg}</Popover.Content>
    </Popover>
  );

  render() {
    const {
      input: { value, ...inputProps },
      className,
      id,
      accept,
      disabled,
      esconderAsterisco,
      helpText,
      label,
      labelClassName,
      meta,
      name,
      placeholder,
      required,
      initialFiles,
      title,
      tipoDocumento,
      tipoArquivo,
      tooltipMessage
    } = this.props;
    return (
      <div className="input input-file-upload">
        {label && [
          required && !esconderAsterisco && (
            <span key={1} className="required-asterisk">
              *
            </span>
          ),
          <label
            key={2}
            htmlFor={name}
            className={`col-form-label ${labelClassName}`}
          >
            {label}
          </label>
        ]}

        {title && [
          <label key={2} htmlFor={name} className={"title"}>
            {title}
          </label>,
          tooltipMessage && (
            <OverlayTrigger
              trigger="click"
              placement="right"
              overlay={this.popover(tooltipMessage)}
            >
              <Image src={imgTooltipQuestion} fluid className="mb-1 ml-1" />
            </OverlayTrigger>
          )
        ]}

        <CustomFileUploadPR
          ref={this.fileUpload}
          disabled={disabled}
          name={name}
          id={id}
          placeholder={placeholder}
          required={required}
          className={`${className} 
             ${meta.touched && meta.error && "invalid-field"}`}
          {...inputProps}
          accept={accept}
          auto={true}
          multiple={true}
          invalidFileSizeMessageSummary={"Erro ao dar upload:"}
          invalidFileSizeMessageDetail={"O tamanho máximo de um arquivo é 15MB"}
          chooseLabel="Selecione os arquivos"
          cancelLabel="Cancelar"
          onUploadChange={this.onChange}
          initialFiles={initialFiles}
          tipoDocumento={tipoDocumento}
          tipoArquivo={tipoArquivo}
        />
        <HelpText helpText={helpText} />
        <InputErroMensagem meta={meta} />
      </div>
    );
  }
}
