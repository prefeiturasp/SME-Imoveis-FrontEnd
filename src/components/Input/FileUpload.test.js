import React from "react";
import { mount } from "enzyme";

import { required } from "helpers/fieldValidators";
import { FileUpload } from "./FileUpload";

describe("test <FileUpload>", () => {
  let wrapper;
  let onChange = jest.fn();
  let input = { onChange, value: "" };

  beforeEach(() => {
    const props = {
      meta: {
        touched: true,
        error: "This field is required"
      }
    };
    wrapper = mount(
      <FileUpload
        className="teste"
        name="myCombo"
        input={input}
        label="mylabel"
        required
        value=""
        validate={required}
        {...props}
      />
    );
  });

  it("renders component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
