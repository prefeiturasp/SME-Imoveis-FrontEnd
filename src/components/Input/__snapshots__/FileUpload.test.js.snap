// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <FileUpload> renders component 1`] = `
<FileUpload
  className="teste"
  input={
    Object {
      "onChange": [MockFunction],
      "value": "",
    }
  }
  label="mylabel"
  meta={
    Object {
      "error": "This field is required",
      "touched": true,
    }
  }
  name="myCombo"
  required={true}
  validate={[Function]}
  value=""
>
  <div
    className="input"
  >
    <span
      className="required-asterisk"
    >
      *
    </span>
    <label
      className="col-form-label undefined"
      htmlFor="myCombo"
      key="1"
    >
      mylabel
    </label>
    <input
      className="form-control teste invalid-field"
      name="myCombo"
      onChange={[Function]}
      required={true}
      type="file"
    />
    <HelpText>
      <div
        className="help-text"
      />
    </HelpText>
    <InputErroMensagem
      meta={
        Object {
          "error": "This field is required",
          "touched": true,
        }
      }
    >
      <div
        className="error-or-warning-message"
      >
        <div
          className="error-message"
        >
          This field is required
        </div>
      </div>
    </InputErroMensagem>
  </div>
</FileUpload>
`;
